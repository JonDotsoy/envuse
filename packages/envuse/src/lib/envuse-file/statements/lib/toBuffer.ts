export function b(data: string | number[]) {
  if (typeof data === "string") return Buffer.from(data);
  return Buffer.from(data);
}

b.c = Buffer.concat;
