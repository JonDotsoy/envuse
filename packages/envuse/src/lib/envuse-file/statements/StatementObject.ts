import { Base } from "./Base";
import { BufferCursor } from "./BufferCursor";
import { BCharType } from "./BCharType";
import { CharactersKey as k } from "./CharactersKey";
import { toBuffer as b, toBuffer } from "./toBuffer";
import { StatementObjectTypes } from "./StatementObjectTypes";
import util from "util";

abstract class StatementObjectDefinition {
  type!: string;
  value: any;

  constructor(readonly rejectUnexpectedTokenError: () => never) {}

  abstract assert(
    bufferCursor: BufferCursor<BCharType>
  ): bufferCursor is BufferCursor<number>;
  abstract prepare(bufferCursor: BufferCursor<BCharType>): void;
}

class StatementTrueObject extends StatementObjectDefinition {
  assert(
    bufferCursor: BufferCursor<BCharType>
  ): bufferCursor is BufferCursor<number> {
    return (
      bufferCursor.has() && b("true").equals(b(bufferCursor.currentAndNext(4)))
    );
  }

  prepare(bufferCursor: BufferCursor<BCharType>): void {
    this.type = StatementObjectTypes.Boolean;
    this.value = true;
    bufferCursor.forward(4);
    return;
  }
}

class StatementFalseObject extends StatementObjectDefinition {
  assert(
    bufferCursor: BufferCursor<BCharType>
  ): bufferCursor is BufferCursor<number> {
    return (
      bufferCursor.has() && b("true").equals(b(bufferCursor.currentAndNext(4)))
    );
  }

  prepare(bufferCursor: BufferCursor<BCharType>): void {
    this.type = StatementObjectTypes.Boolean;
    this.value = false;
    bufferCursor.forward(5);
    return;
  }
}

class StatementStrictEqualObject extends StatementObjectDefinition {
  assert(
    bufferCursor: BufferCursor<BCharType>
  ): bufferCursor is BufferCursor<number> {
    return (
      bufferCursor.has() && b("===").equals(b(bufferCursor.currentAndNext(3)))
    );
  }

  prepare(bufferCursor: BufferCursor<BCharType>): void {
    this.type = StatementObjectTypes.StrictEquality;
    bufferCursor.forward(3);
    return;
  }
}

class StatementNameInstanceObject extends StatementObjectDefinition {
  assert(
    bufferCursor: BufferCursor<BCharType>
  ): bufferCursor is BufferCursor<number> {
    return (
      bufferCursor.has() && k.english_alphabet.includes(bufferCursor.current())
    );
  }

  prepare(bufferCursor: BufferCursor<number>): void {
    this.type = StatementObjectTypes.NameInstance;
    let partialPath: string[] = [];
    let charAccumulation: number[] = [];

    while (true) {
      if (
        bufferCursor.isClosed() ||
        bufferCursor.current() === k.space ||
        bufferCursor.current() === k.newLineLF
      ) {
        partialPath.push(toBuffer(charAccumulation).toString());
        this.value = partialPath;
        // bufferCursor.forward();
        return;
      }

      if (bufferCursor.current() === k.dot) {
        partialPath.push(toBuffer(charAccumulation).toString());
        charAccumulation = [];
        bufferCursor.forward();
        continue;
      }

      if (
        k.english_alphabet.includes(bufferCursor.current()) ||
        k.underscore === bufferCursor.current() ||
        k.numbers.includes(bufferCursor.current())
      ) {
        charAccumulation.push(bufferCursor.current());
        bufferCursor.forward();
        continue;
      }

      this.rejectUnexpectedTokenError();
    }
  }
}

class StatementStringObject extends StatementObjectDefinition {
  assert(
    bufferCursor: BufferCursor<BCharType>
  ): bufferCursor is BufferCursor<number> {
    return (
      bufferCursor.has() &&
      (k.doubleQuotes === bufferCursor.current() ||
        k.singleQuote === bufferCursor.current())
    );
  }

  prepare(bufferCursor: BufferCursor<number>): void {
    const raw: number[] = [];
    const appendRaw = (...chars: number[]) => {
      raw.push(...chars);
    };

    this.type = StatementObjectTypes.String;
    const initialQuote = bufferCursor.current();
    bufferCursor.forward();

    while (true) {
      const p = bufferCursor.prev(2);
      if (
        bufferCursor.isClosed() ||
        initialQuote === bufferCursor.current() ||
        bufferCursor.current() === k.newLineLF
      ) {
        bufferCursor.forward();
        this.value = String(b(raw));
        return;
      }

      if (bufferCursor.current() == k.backslash) {
        bufferCursor.forward();
        switch (bufferCursor.current()) {
          case 0x74: {
            appendRaw(k.horizontalTab);
            bufferCursor.forward();
            continue;
          }
          case 0x6e: {
            appendRaw(k.newLineLF);
            bufferCursor.forward();
            continue;
          }
          default: {
            appendRaw(bufferCursor.current());
            bufferCursor.forward();
            continue;
          }
        }
      }

      appendRaw(bufferCursor.current());
      bufferCursor.forward();
    }
  }
}

class StatementNumberObject extends StatementObjectDefinition {
  assert(
    bufferCursor: BufferCursor<BCharType>
  ): bufferCursor is BufferCursor<number> {
    return bufferCursor.has() && k.numbers.includes(bufferCursor.current());
  }

  prepare(bufferCursor: BufferCursor<number>): void {
    const raw: number[] = [];
    const appendRaw = (...chars: number[]) => {
      raw.push(...chars);
    };

    this.type = StatementObjectTypes.Number;
    while (true) {
      if (
        bufferCursor.isClosed() ||
        k.space === bufferCursor.current() ||
        bufferCursor.current() === k.newLineLF ||
        bufferCursor.current() === k.equalsSign
      ) {
        this.value = Number(b(raw));
        return;
      }
      if (
        k.numbers.includes(bufferCursor.current()) ||
        bufferCursor.current() === k.dot
      ) {
        appendRaw(bufferCursor.current());
        bufferCursor.forward();
        continue;
      }
      if (bufferCursor.current() === k.underscore) {
        bufferCursor.forward();
        continue;
      }
      this.rejectUnexpectedTokenError();
    }
  }
}

export class StatementObject extends Base {
  static types = StatementObjectTypes;

  type!: string;
  value: any;

  definitions: StatementObjectDefinition[] = [
    new StatementTrueObject(() => this.rejectUnexpectedTokenError()),
    new StatementFalseObject(() => this.rejectUnexpectedTokenError()),
    new StatementStrictEqualObject(() => this.rejectUnexpectedTokenError()),
    new StatementNameInstanceObject(() => this.rejectUnexpectedTokenError()),
    new StatementStringObject(() => this.rejectUnexpectedTokenError()),
    new StatementNumberObject(() => this.rejectUnexpectedTokenError()),
  ];

  toObjectName() {
    return `${this.constructor.name}<${this.type ?? "unknown"}>`;
  }

  prepare(bufferCursor: BufferCursor<BCharType>): void {
    if (bufferCursor.has()) {
      for (const definition of this.definitions) {
        if (definition.assert(bufferCursor)) {
          definition.prepare(bufferCursor);
          this.type = definition.type;
          this.value = definition.value;
          return;
        }
      }
    }

    this.rejectUnexpectedTokenError();
  }
}
