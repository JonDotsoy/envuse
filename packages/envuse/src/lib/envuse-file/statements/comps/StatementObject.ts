import { Base, BaseExportTypeJSON } from "./Base";
import { BufferCursor } from "../lib/BufferCursor";
import { BCharType } from "../tdo/BCharType";
import { StatementObjectTypes } from "../tdo/StatementObjectTypes";
import { typesValues } from "./typesValues";

import { StatementTrueObject } from "./statement-object/StatementTrueObject";
import { StatementFalseObject } from "./statement-object/StatementFalseObject";
import { StatementStrictEqualObject } from "./statement-object/StatementStrictEqualObject";
import { StatementNameInstanceObject } from "./statement-object/StatementNameInstanceObject";
import { StatementStringObject } from "./statement-object/StatementStringObject";
import { StatementNumberObject } from "./statement-object/StatementNumberObject";


type StatementObjectTypeBase =
  | {
    $type: 'StatementObject'
    type: 'NameInstance'
    value: string[]
  }
  | {
    $type: 'StatementObject'
    type: 'String'
    value: string
  }
  | {
    $type: 'StatementObject'
    type: 'Boolean'
    value: boolean
  }
  | {
    $type: 'StatementObject'
    type: 'Number'
    value: number
  }
  | {
    $type: 'StatementObject'
    type: 'StrictEqualitySymbol'
    value: '==='
  }

export type StatementObjectType = StatementObjectTypeBase & { [k: string]: any }


export class StatementObject extends Base {
  $type = 'StatementObject' as const;

  static types = StatementObjectTypes;

  type!: typesValues;
  value: any;

  readonly definitions = [
    new StatementTrueObject(() => this.rejectUnexpectedTokenError()),
    new StatementFalseObject(() => this.rejectUnexpectedTokenError()),
    new StatementStrictEqualObject(() => this.rejectUnexpectedTokenError()),
    new StatementNameInstanceObject(() => this.rejectUnexpectedTokenError()),
    new StatementStringObject(() => this.rejectUnexpectedTokenError()),
    new StatementNumberObject(() => this.rejectUnexpectedTokenError()),
  ] as const;

  toObjectName() {
    return `${this.constructor.name}<${this.type ?? "unknown"}>`;
  }

  prepare(bufferCursor: BufferCursor<BCharType>): void {
    if (bufferCursor.has()) {
      for (const definition of this.definitions) {
        if (definition.assert(bufferCursor)) {
          definition.prepare(bufferCursor);
          this.type = definition.type;
          this.value = definition.value;
          return;
        }
      }
    }

    this.rejectUnexpectedTokenError();
  }

  toJSON() {
    return <BaseExportTypeJSON<'StatementObject'> & StatementObjectType>{
      ...super.toJSON(),
      type: this.type,
      value: this.value
    }
  }

  static serialize(comp: StatementObjectType) {
    switch (comp.type) {
      case 'NameInstance':
        return Buffer.from(comp.value.join('.'))
      case 'StrictEqualitySymbol':
        return Buffer.from('===')
      case 'String':
        return Buffer.from(JSON.stringify(comp.value.toString()))
      case 'Number':
      case 'Boolean':
        return Buffer.from(comp.value.toString())
      default: {
        // @ts-ignore
        throw new Error(`type ${comp.type} is not supported`)
      }
    }
  }
}
