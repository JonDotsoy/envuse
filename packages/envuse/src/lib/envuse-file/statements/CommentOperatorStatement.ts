import { Base } from "./Base";
import { BufferCursor } from "./BufferCursor";
import { BCharType } from "./BCharType";
import { Space } from "./Space";
import { CharactersKey as K } from "./CharactersKey";
import { StatementObject } from "./StatementObject";


export class CommentOperatorStatement extends Base {
  prepare(bufferCursor: BufferCursor<BCharType>): void {
    while (true) {
      if (bufferCursor.current() === K.newLineLF || bufferCursor.current() === K.numberSign || bufferCursor.isClosed()) {
        bufferCursor.forward();
        return;
      }

      this.children.push(this.createElement(StatementObject));

      if (bufferCursor.current() === K.newLineLF) {
        bufferCursor.forward()
        return
      }

      if (bufferCursor.current() === K.space) {
        this.createElement(Space);
      }

    }
  }
}
