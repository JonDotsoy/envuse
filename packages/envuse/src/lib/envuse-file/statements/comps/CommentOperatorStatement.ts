import { Base } from "./Base";
import { BufferCursor } from "../lib/BufferCursor";
import { BCharType } from "../tdo/BCharType";
import { Space } from "./Space";
import { CharactersKey as K } from "../tdo/CharactersKey";
import { StatementObject } from "./StatementObject";

export class CommentOperatorStatement extends Base {
  $type = 'CommentOperatorStatement' as const;
  statements: StatementObject[] = []

  prepare(bufferCursor: BufferCursor<BCharType>): void {
    while (true) {
      if (
        bufferCursor.current() === K.newLineLF ||
        bufferCursor.current() === K.numberSign ||
        bufferCursor.isClosed()
      ) {
        bufferCursor.forward();
        return;
      }

      this.statements.push(this.createElement(StatementObject));

      if (bufferCursor.current() === K.newLineLF) {
        bufferCursor.forward();
        return;
      }

      if (bufferCursor.current() === K.space) {
        this.createElement(Space);
      }
    }
  }

  toJSON() {
    return {
      ...super.toJSON(),
      statements: this.statements,
    }
  }
}
