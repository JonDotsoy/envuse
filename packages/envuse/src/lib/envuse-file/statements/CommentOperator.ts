import { Base } from "./Base";
import { BufferCursor } from "./BufferCursor";
import { BCharType } from "./BCharType";
import { Space } from "./Space";
import { VariableKey } from "./Variable";
import { Block } from "./Block";
import { CharactersKey as K } from "./CharactersKey";
import { toBuffer as b } from "./toBuffer";
import { CommentOperatorStatement } from "./CommentOperatorStatement";

export class CommentOperator extends Base {
  operator!: VariableKey;
  statement!: CommentOperatorStatement;
  block!: Block;

  prepare(bufferCursor: BufferCursor<BCharType>): void {
    bufferCursor.forward(2);

    if (bufferCursor.current() === K.space) this.createElement(Space);

    this.operator = this.createElement(VariableKey);

    switch (this.operator.raw.toString()) {
      case "if":
        break;
      case "fi":
        break;
      default: {
        bufferCursor.backward(this.operator.raw.length);
        this.rejectUnexpectedTokenError();
      }
    }

    if (this.operator.raw.equals(b("fi"))) {
      return;
    }

    this.createElement(Space);

    this.statement = this.createElement(CommentOperatorStatement);

    this.block = this.createElement(Block, {
      handleCheckCloseBlock() {
        const lastChild = this.children[this.children.length - 1];
        if (
          lastChild instanceof CommentOperator &&
          lastChild.operator.raw.equals(b("fi"))
        ) {
          return true;
        }
        return false;
      },
    });
  }
}
