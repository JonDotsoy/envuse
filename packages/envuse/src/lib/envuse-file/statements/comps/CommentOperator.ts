import { Base } from "./Base";
import { BufferCursor } from "../lib/BufferCursor";
import { BCharType } from "../tdo/BCharType";
import { Space } from "./Space";
import { VariableKey } from "./VariableKey";
import { Block } from "./Block";
import { CharactersKey as K } from "../tdo/CharactersKey";
import { toBuffer as b } from "../lib/toBuffer";
import { CommentOperatorStatement } from "./CommentOperatorStatement";


export class CommentOperator extends Base {
  $type = 'CommentOperator' as const;
  operator!: VariableKey;
  statement!: CommentOperatorStatement;
  block!: Block;

  prepare(bufferCursor: BufferCursor<BCharType>): void {
    bufferCursor.forward(2);

    if (bufferCursor.current() === K.space) this.createElement(Space);

    this.operator = this.createElement(VariableKey);

    switch (this.operator.raw.toString()) {
      case "if":
        break;
      case "fi":
        break;
      default: {
        bufferCursor.backward(this.operator.raw.length);
        this.rejectUnexpectedTokenError();
      }
    }

    if (this.operator.raw.equals(b("fi"))) {
      return;
    }

    this.createElement(Space);

    this.statement = this.createElement(CommentOperatorStatement);

    this.block = this.createElement(Block, {
      handleCheckCloseBlock() {
        const lastChild = this.children[this.children.length - 1];
        if (
          lastChild instanceof CommentOperator &&
          lastChild.operator.raw.equals(b("fi"))
        ) {
          return true;
        }
        return false;
      },
    });
  }

  toJSON() {
    return {
      ...super.toJSON(),
      operator: this.operator,
      statement: this.statement,
      block: this.block,
    }
  }
}
