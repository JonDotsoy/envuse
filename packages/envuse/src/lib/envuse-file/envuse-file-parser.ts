import { eventNames } from "process";
import { Base } from "./statements/Base";
import { Block } from "./statements/Block";
import { ArrCursor } from "./statements/ArrCursor";
import { CommentOperator } from "./statements/CommentOperator";
import { UnexpectedTokenError } from "./statements/UnexpectedTokenError";
import { Variable } from "./statements/Variable";
import { StatementObject } from "./statements/StatementObject";

type Option = {
  filename?: string | null;
  body: Buffer;
};

/** AST Parser */
export class EnvuseFileParser {
  constructor(private filename: string | null, private body: Buffer) { }

  toAstBody() {
    try {
      return Base.createElement(new Block(this.filename, this.body, 0));
    } catch (err) {
      if (UnexpectedTokenError.isUnexpectedTokenError(err)) {
        // Error.captureStackTrace(err, EnvuseFileParser.prototype.toAstBody);
        // const stack = err.stack;
        // if (stack) {
        //   // let posEndMessage = err.name.length + err.message.length + 3;
        // }
      }
      throw err;
    }
  }

  private static makeOperatorEvaluator(operator: CommentOperator) {
    return (value: { [k: string]: any }) => {
      const a = new ArrCursor(operator.statement.statements)
      let accumulator: any;

      while (a.has()) {
        if (a.position === 0) {
          accumulator = a.current().value
          a.forward()
          continue
        }

        switch (a.current().type) {
          case 'StrictEqualitySymbol': {
            a.forward()
            accumulator = accumulator === a.current().value
            a.forward()
            break
          }
          default: throw new Error(`Statement no expected`)
        }
      }

      return accumulator
    }
  }

  static parse(options: Option) {
    const ast = this.parseToAst(options)

    const operatorsList = ast.elementList
      .filter((element): element is CommentOperator => element instanceof CommentOperator)
      .map((operator) => ({
        operator,
        assert: this.makeOperatorEvaluator(operator),
      }))

    const variableList = ast.elementList.filter((element): element is Variable => element instanceof Variable)

    const parsed = variableList
      .reduce((acum, element) => {
        const operators = operatorsList.filter(operator => operator.operator.elementList.includes(element))
        if (operators.length) {
          const result = operators.reduce((v, operator) => v && operator.assert(acum), true)
          if (!result) {
            return acum;
          }
        }
        return {
          ...acum,
          [element.keyVariable.value]: element.valueVariable.value
        };
      }, {} as { [k: string]: any; });

    return { parsed } as const
  }

  static parseToAst(options: Option) {
    return new EnvuseFileParser(options.filename ?? null, options.body).toAstBody();
  }
}
