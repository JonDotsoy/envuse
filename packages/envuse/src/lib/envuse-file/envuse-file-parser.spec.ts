import { EnvuseFileParser } from "./envuse-file-parser";
import { takeDemoFile } from "./statements/takeDemoFile";
import util from 'util'
import { CommentOperator } from "./statements/CommentOperator";
import { CommentOperatorStatement } from "./statements/CommentOperatorStatement";
import { StatementObject } from "./statements/StatementObject";
import { Base, printElement } from "./statements/Base";
import { toBuffer as b } from "./statements/toBuffer";

describe('EnvuseFileParser2', () => {

  it('shoud make a ast out', () => {
    const [filename1, body1] = takeDemoFile('.env');

    const envuseFileParser = new EnvuseFileParser(filename1, body1);
    expect(util.formatWithOptions({ depth: Infinity }, '', envuseFileParser.toAstBody())).toMatchSnapshot()
  })

  it('shoud make a ast with operators', () => {
    const [filename, body] = takeDemoFile('.env');

    const envuseFileParser = new EnvuseFileParser(filename, body);
    expect(util.formatWithOptions({ depth: Infinity }, '', envuseFileParser.toAstBody())).toMatchSnapshot()
    // console.log(util.formatWithOptions({ depth: Infinity }, '', envuseFileParser.toAstBody()))
  })

  it('shoud make a ast with multiple blocks', () => {
    const [filename, body] = takeDemoFile('.env');

    const envuseFileParser = new EnvuseFileParser(filename, body);
    expect(util.formatWithOptions({ depth: Infinity }, '', envuseFileParser.toAstBody())).toMatchSnapshot()
    // console.log(util.formatWithOptions({ depth: Infinity }, '', envuseFileParser.toAstBody()))
  })

  describe('parse statement', () => {

    it('parse statement operation', () => {
      const body = Buffer.from('true ===   false # asd');

      const cmp = Base.createElement(new CommentOperatorStatement('a', body, 0))

      expect(cmp).toMatchSnapshot()
    })

    it('should parse statement with numbers', () => {
      const body = Buffer.from('34214324 ===   22_343.32 # comment');

      const cmp = Base.createElement(new CommentOperatorStatement('a', body, 0))

      expect(cmp).toMatchSnapshot()
      // console.log(cmp)
    })

    it('should parse StatementObject string', () => {
      const [fl, body] = takeDemoFile()

      const cmp = Base.createElement(new StatementObject(fl, body, 0))

      // expect(cmp).toMatchSnapshot()
      // console.log(cmp)
    })

    it('should parse statement with string', () => {
      const [fl, body] = takeDemoFile()

      Base.createElement(new CommentOperatorStatement(fl, body, 0))

      // expect(cmp).toMatchSnapshot()
      // console.log(cmp)
    })

    it('should parse statement with string and special characters', () => {
      const [fl, body] = takeDemoFile('.special-characters.txt')

      const cmp = Base.createElement(new StatementObject(fl, body, 0))

      expect(cmp.value).toEqual('\n\t')
    })

    it('should parse instance path', () => {
      const body = Buffer.from('obj.a.b.VAL2');

      const cmp = Base.createElement(new StatementObject('a', body, 0))

      expect(cmp.value).toEqual(['obj', 'a', 'b', 'VAL2'])
    })

    it('should parse numbers', () => {
      const body = b('1_321.123');

      const cmp = Base.createElement(new StatementObject('a', body, 0))

      expect(cmp.value).toEqual(1_321.123)
    })

    it('should parse strings', () => {
      const body = b('"hi"');

      const cmp = Base.createElement(new StatementObject(null, body, 0))

      expect(cmp.value).toEqual("hi")
    })

    it('should parse statement with variables', () => {
      const body = Buffer.from('VAL1 ===   obj.a.b.VAL2 # comment');

      const cmp = Base.createElement(new CommentOperatorStatement('a', body, 0))

      expect(cmp.elementList.map((e) => e.toString())).toMatchInlineSnapshot(`
        Array [
          "CommentOperatorStatement (0, 25): VAL1 ===   obj.a.b.VAL2 #",
          "StatementObject<NameInstance> (0, 4): VAL1",
          "Space (4, 5):  ",
          "StatementObject<StrictEquality> (5, 8): ===",
          "Space (8, 11):    ",
          "StatementObject<NameInstance> (11, 23): obj.a.b.VAL2",
          "Space (23, 24):  ",
        ]
      `)
    })

    it('should parse end parsing comment code', () => {
      const body = b('#! if true\nFOO=bar')

      const cmp = Base.createElement(new CommentOperator('a', body, 0))

      // expect(cmp).toMatchInlineSnapshot(`Object {}`)
      expect(cmp.operator.raw.toString()).toEqual('if')
      // console.log(cmp)
    })

    it('should read property list children', () => {
      const [fl, body] = takeDemoFile('.env')

      const envuseFileParser = new EnvuseFileParser(fl, body);

      const b = envuseFileParser.toAstBody()

      for (const elem of b.elementList) {
        expect(elem).toBeInstanceOf(Base)
      }

      expect(b.elementList.map((e) => `${e.constructor.name} <${e.pos}, ${e.end}>`)).toMatchInlineSnapshot(`
        Array [
          "Block <0, 35>",
          "Variable <0, 8>",
          "VariableKey <0, 3>",
          "SymbolEqual <3, 4>",
          "VariableValue <4, 8>",
          "CommentOperator <8, 34>",
          "Space <10, 11>",
          "VariableKey <11, 13>",
          "Space <13, 14>",
          "CommentOperatorStatement <14, 19>",
          "StatementObject <14, 18>",
          "Block <19, 34>",
          "SpaceNewLine <19, 20>",
          "Variable <20, 28>",
          "VariableKey <20, 23>",
          "SymbolEqual <23, 24>",
          "VariableValue <24, 28>",
          "SpaceNewLine <28, 29>",
          "CommentOperator <29, 34>",
          "Space <31, 32>",
          "VariableKey <32, 34>",
          "SpaceNewLine <34, 35>",
        ]
      `);
    })

    it('should parse comment operator correctly', () => {
      expect(EnvuseFileParser.parse({ filename: null, body: Buffer.from(`#;if a.b.c.d ===\nfoo="bar"\n#;fi\n`) }).elementList.map(e => e.toString())).toMatchSnapshot()
      expect(EnvuseFileParser.parse({ filename: null, body: Buffer.from(`#;if true\nfoo="bar"\n#;fi\n`) }).elementList.map(e => e.toString())).toMatchSnapshot()
      expect(EnvuseFileParser.parse({ filename: null, body: Buffer.from(`#;if 123\nfoo="bar"\n#;fi\n`) }).elementList.map(e => e.toString())).toMatchSnapshot()
      expect(EnvuseFileParser.parse({ filename: null, body: Buffer.from(`#;if var\nfoo="bar"\n#;fi\n`) }).elementList.map(e => e.toString())).toMatchSnapshot()
      expect(EnvuseFileParser.parse({ filename: null, body: Buffer.from(`#;if foo === bar\nfoo="bar"\n#;fi\n`) }).elementList.map(e => e.toString())).toMatchSnapshot()
    })

  })

})
