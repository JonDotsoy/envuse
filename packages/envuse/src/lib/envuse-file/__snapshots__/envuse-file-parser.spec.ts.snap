// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EnvuseFileParser2 parse statement parse statement operation 1`] = `Object {}`;

exports[`EnvuseFileParser2 parse statement should parse comment operator correctly 1`] = `
Array [
  "Block (0, 32): #;if a.b.c.d ===
foo=\\"bar\\"
#;fi
",
  "CommentOperator (0, 31): #;if a.b.c.d ===
foo=\\"bar\\"
#;fi",
  "VariableKey (2, 4): if",
  "Space (4, 5):  ",
  "CommentOperatorStatement (5, 17): a.b.c.d ===
",
  "StatementObject<NameInstance> (5, 12): a.b.c.d",
  "Space (12, 13):  ",
  "StatementObject<StrictEquality> (13, 16): ===",
  "Block (17, 31): foo=\\"bar\\"
#;fi",
  "Variable (17, 26): foo=\\"bar\\"",
  "VariableKey (17, 20): foo",
  "SymbolEqual (20, 21): =",
  "VariableValue (21, 26): \\"bar\\"",
  "SpaceNewLine (26, 27): 
",
  "CommentOperator (27, 31): #;fi",
  "VariableKey (29, 31): fi",
  "SpaceNewLine (31, 32): 
",
]
`;

exports[`EnvuseFileParser2 parse statement should parse comment operator correctly 2`] = `
Array [
  "Block (0, 25): #;if true
foo=\\"bar\\"
#;fi
",
  "CommentOperator (0, 24): #;if true
foo=\\"bar\\"
#;fi",
  "VariableKey (2, 4): if",
  "Space (4, 5):  ",
  "CommentOperatorStatement (5, 10): true
",
  "StatementObject<Boolean> (5, 9): true",
  "Block (10, 24): foo=\\"bar\\"
#;fi",
  "Variable (10, 19): foo=\\"bar\\"",
  "VariableKey (10, 13): foo",
  "SymbolEqual (13, 14): =",
  "VariableValue (14, 19): \\"bar\\"",
  "SpaceNewLine (19, 20): 
",
  "CommentOperator (20, 24): #;fi",
  "VariableKey (22, 24): fi",
  "SpaceNewLine (24, 25): 
",
]
`;

exports[`EnvuseFileParser2 parse statement should parse comment operator correctly 3`] = `
Array [
  "Block (0, 24): #;if 123
foo=\\"bar\\"
#;fi
",
  "CommentOperator (0, 23): #;if 123
foo=\\"bar\\"
#;fi",
  "VariableKey (2, 4): if",
  "Space (4, 5):  ",
  "CommentOperatorStatement (5, 9): 123
",
  "StatementObject<Number> (5, 8): 123",
  "Block (9, 23): foo=\\"bar\\"
#;fi",
  "Variable (9, 18): foo=\\"bar\\"",
  "VariableKey (9, 12): foo",
  "SymbolEqual (12, 13): =",
  "VariableValue (13, 18): \\"bar\\"",
  "SpaceNewLine (18, 19): 
",
  "CommentOperator (19, 23): #;fi",
  "VariableKey (21, 23): fi",
  "SpaceNewLine (23, 24): 
",
]
`;

exports[`EnvuseFileParser2 parse statement should parse comment operator correctly 4`] = `
Array [
  "Block (0, 24): #;if var
foo=\\"bar\\"
#;fi
",
  "CommentOperator (0, 23): #;if var
foo=\\"bar\\"
#;fi",
  "VariableKey (2, 4): if",
  "Space (4, 5):  ",
  "CommentOperatorStatement (5, 9): var
",
  "StatementObject<NameInstance> (5, 8): var",
  "Block (9, 23): foo=\\"bar\\"
#;fi",
  "Variable (9, 18): foo=\\"bar\\"",
  "VariableKey (9, 12): foo",
  "SymbolEqual (12, 13): =",
  "VariableValue (13, 18): \\"bar\\"",
  "SpaceNewLine (18, 19): 
",
  "CommentOperator (19, 23): #;fi",
  "VariableKey (21, 23): fi",
  "SpaceNewLine (23, 24): 
",
]
`;

exports[`EnvuseFileParser2 parse statement should parse comment operator correctly 5`] = `
Array [
  "Block (0, 32): #;if foo === bar
foo=\\"bar\\"
#;fi
",
  "CommentOperator (0, 31): #;if foo === bar
foo=\\"bar\\"
#;fi",
  "VariableKey (2, 4): if",
  "Space (4, 5):  ",
  "CommentOperatorStatement (5, 17): foo === bar
",
  "StatementObject<NameInstance> (5, 8): foo",
  "Space (8, 9):  ",
  "StatementObject<StrictEquality> (9, 12): ===",
  "Space (12, 13):  ",
  "StatementObject<NameInstance> (13, 16): bar",
  "Block (17, 31): foo=\\"bar\\"
#;fi",
  "Variable (17, 26): foo=\\"bar\\"",
  "VariableKey (17, 20): foo",
  "SymbolEqual (20, 21): =",
  "VariableValue (21, 26): \\"bar\\"",
  "SpaceNewLine (26, 27): 
",
  "CommentOperator (27, 31): #;fi",
  "VariableKey (29, 31): fi",
  "SpaceNewLine (31, 32): 
",
]
`;

exports[`EnvuseFileParser2 parse statement should parse statement with numbers 1`] = `Object {}`;

exports[`EnvuseFileParser2 shoud make a ast out 1`] = `
" Block (0, 62):  abc=def # a comment
foo=bar

    
abc    =\\"d\\\\\\"ef\\" # comented
"
`;

exports[`EnvuseFileParser2 shoud make a ast with multiple blocks 1`] = `
" Block (0, 42): C=C
#;if true
a=a
#;if false
b=b
#;fi
#;fi"
`;

exports[`EnvuseFileParser2 shoud make a ast with operators 1`] = `
" Block (0, 43): FOO=BAR
#; if true
aaa=\\"bbb\\"
#; fi

ccc=ddd"
`;
